be/src/storage/persistent_index.cpp

```diff
@@ -25,6 +72,378 @@ struct FixedKeyHash {
+            auto bucket_packs = bucket_packs_in_page[i] + bucket_packs_in_page[j];
+            if (bucket_packs == min_pack_to_move) {
+                buckets_to_move->emplace_back(bucket_packs_in_page[i], pageid, i);
+                buckets_to_move->emplace_back(bucket_packs_in_page[j], pageid, j);
+                return Status::OK();
+            } else if (bucket_packs > min_pack_to_move && bucket_packs < move2_min[0]) {
+                move2_min = {bucket_packs, i, j};
+                break;
+            }
+        }
+    }
+    if (move2_min[2] >= 0) {
+        for (int i = 1; i < 3; i++) {
+            auto bucketid = move2_min[i];
+            buckets_to_move->emplace_back(bucket_packs_in_page[bucketid], pageid, bucketid);
+        }
+        return Status::OK();
+    }
+    // try find solution of moving 3 bucket
+    std::array<int, 4> move3_min = {INT_MAX, -1, -1, -1};
+    for (int oi = 0; oi < bucket_per_page; oi++) {
+        int i = bucketid_ordered_by_packs[oi];
+        if (bucket_packs_in_page[i] == 0) {
+            continue;
+        }
+        for (int oj = oi + 1; oj < bucket_per_page; oj++) {
+            int j = bucketid_ordered_by_packs[oj];
+            for (int ok = oj + 1; ok < bucket_per_page; ok++) {
+                int k = bucketid_ordered_by_packs[ok];
+                auto bucket_packs = bucket_packs_in_page[i] + bucket_packs_in_page[j] + bucket_packs_in_page[k];
+                if (bucket_packs == min_pack_to_move) {
+                    buckets_to_move->emplace_back(bucket_packs_in_page[i], pageid, i);
+                    buckets_to_move->emplace_back(bucket_packs_in_page[j], pageid, j);
+                    buckets_to_move->emplace_back(bucket_packs_in_page[k], pageid, k);
+                    return Status::OK();
+                } else if (bucket_packs > min_pack_to_move && bucket_packs < move3_min[0]) {
+                    move3_min = {bucket_packs, i, j, k};
+                    break;
+                }
+            }
+        }
+    }
+    if (move3_min[2] >= 0) {
+        for (int i = 1; i < 4; i++) {
+            auto bucketid = move3_min[i];
+            buckets_to_move->emplace_back(bucket_packs_in_page[bucketid], pageid, bucketid);
+        }
+        return Status::OK();
+    }
+    // TODO: current algorithm is sub-optimal, find buckets to move using DP
+    return Status::InternalError("find_buckets_to_move");
+}
+
+struct BucketMovement {
+    uint32_t src_pageid;
+    uint32_t src_bucketid;
+    uint32_t dest_pageid;
+    BucketMovement(uint32_t src_pageid, uint32_t src_bucketid, uint32_t dest_pageid)
+            : src_pageid(src_pageid), src_bucketid(src_bucketid), dest_pageid(dest_pageid) {}
+};
+
+static void remove_packs_from_dests(std::vector<MoveDest>& dests, int idx, int npack) {
+    auto& d = dests[idx];
+    d.npack -= npack;
+    if (d.npack == 0) {
+        dests.erase(dests.begin() + idx);
+    } else {
+        for (int c = idx; c > 0; c--) {
```

Comment:
Looks like it could be optimized to two steps, first swap with the lower bound, second shift the range to right?

Reply:
This code is not the bottleneck, why complicate the code.

Reply:
fixed